File: merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        ListNode h1 = new ListNode(-1);
        h1.next = l1;
        ListNode h2 = new ListNode(-1);
        h2.next = l2;
        ListNode r = new ListNode(-1);
        ListNode t = r;
        ListNode t1 = l1;
        ListNode t2 = l2;
        while (t1 != null && t2 != null) {
            if (t1.val < t2.val) {
                t.next = t1;
                t = t1;
                t1 = t1.next;
            } else {
                t.next = t2;
                t = t2;
                t2 = t2.next;
            }
        }
        while (t1 != null) {
            t.next = t1;
            t = t1;
            t1 = t1.next;
        }
        while(t2 != null) {
            t.next = t2;
            t = t2;
            t2 = t2.next;
        }
        return r.next;
    }
    public ListNode mergeKLists(List<ListNode> lists) {
        if (lists == null || lists.size() == 0) return null;
        if (lists.size() == 1) return lists.get(0);
        int mid = (lists.size() - 1) / 2;
        ListNode l1 = mergeKLists(lists.subList(0, mid+1));
        ListNode l2 = mergeKLists(lists.subList(mid+1, lists.size()));
        return mergeTwoLists(l1, l2);
    }
    public ListNode mergeKLists(ListNode[] lists) {
        List<ListNode> lists1 = Arrays.asList(lists);
        return mergeKLists(lists1);
        
    }
}

File: merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode*h = new ListNode(-1);
        ListNode*tmp = h;
        for (;l1 && l2;) {
            if (l1 -> val < l2 -> val) {
                tmp -> next = l1;
                tmp = l1;
                l1 = l1 -> next;
            } else {
                tmp -> next = l2;
                tmp = l2;
                l2 = l2 -> next;
            }
        }
        if (l1) tmp -> next = l1;
        if (l2) tmp -> next = l2;
        return h -> next;
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return NULL;
        if (lists.size() == 1) {
            return lists[0];
        }
        int mid = lists.size() / 2;
        vector<ListNode*> lists1;for (int i = 0; i < mid; ++i) lists1.push_back(lists[i]);
        vector<ListNode*> lists2;for (int i = mid; i < lists.size(); ++i) lists2.push_back(lists[i]);
        ListNode* l1 = mergeKLists(lists1);
        ListNode* l2 = mergeKLists(lists2);
        return mergeTwoLists(l1, l2);
    }
};